### PR TITLE
General input cleanup + Controller status repolling + Skip polling empty ports + Player numbers + Fix rumble + GC controller bugfixes + Rumble pak graphic translations + Demo fixes + More

### PR DESCRIPTION
**Pull request by Arceveti**
image

Changes/fixes:
- 
- General controller/input:
    - 
    - Everything that's controlled by Mario (including rumble) now uses `gMarioState->controller` when available instead of specifically only player 1 controller
        - Menus still use the player 1 controller.
    - Crash screen now accepts input from any controller
    - Goddard/Mario head now uses the same input polling functionality as the rest of the game
    - The demo now uses a separate virtual controller, so there is no longer a possibility of player input and demo input interfering with each other
    - Added documentation for menu scrolling handling
    - Moved/renamed input stuff into separate files:
        - `input.c` (stuff previously in `game_init.c`)
        - `joybus.c` (previously `gamecube_controller.c`)
        - `rumble.c` (previously `rumble_init.c`)
        - `controller_select_menu.c` (the status repolling/player select menu)
    - Added a `MAX_NUM_PLAYERS` define
    - On boot:
        - If `MAX_NUM_PLAYERS` == 1`, the first controller that has input will be assigned as player 1
        - Otherwise, controllers will be assigned via port order like vanilla

- GameCube controllers:
    - 
    - GameCube controllers' analog origins are now recalibrated properly when holding `X+Y+START` for 3 seconds by sending the "read origin" command
    - Cleaned up converting GCN inputs to N64 inputs
        - See `__osContReadGCNInputData` in `joybus.c`
    - Added separate deadzones for GCN controllers
        - See `adjust_analog_stick` in `input.c`

- Rumble:
    - 
    - Fixed `ENABLE_RUMBLE` ifdefs
    - Documentation for `rumble_init.c`
    - Added translations for the rumble pak graphic
        - TODO: Multilang PR to implement these
    - Only check for rumble pak status once per second (560us-680us saved when a rumble pak is plugged in and not rumbling)
    - It is now possible to use rumble on multiple controllers at once
    - `queue_rumble_data` now specifies a controller and a decay arg
    - Rumble pak can now be plugged/unplugged on any controller at any time and will still work

- Joybus:
    - 
    - Rewritten Joybus functions (see `joybus.c`):
        - `osContReadData` and `osContGetReadData` to a more generalized `osContStartReadDataEx` and `osContGetReadDataEx`
            - Writes commands depending on controller type
            - Skips empty ports and unused controllers
                - This, combined with new `osContSetCh` usage, saves at least 210us-250us saved per skipped controller
        - Rumble/motor functions now work with GCN controllers
        - Documentation for modified functions
    - Added data/documentation for almost all N64/GCN peripherals' protocols

- Controller management/status repolling menu:
    - 
    - This allows the player to
        - See which controllers are currently plugged in (icons for almost every detectable controller type)
        - Plug/unplug any controller (status poll time is controlled by `CONT_STATUS_POLLING_TIME` in `input,h`
        - Set the player number for each controller
    - This is triggered either by:
        - Unplugging any active controller
        - Doing `Z+R+A+START` (N64) or `L+R+A+START` (GCN)
            - The combo is defined as `TOGGLE_CONT_STATUS_POLLING_COMBO` in `input.h`
            - Whether the combo can be used during normal gameplay is controlled by the `ENABLE_STATUS_REPOLLING_COMBO_IN_GAMEPLAY` define in `config_rom.h`
    - Disable `ENABLE_STATUS_REPOLLING_GUI` in `config_rom.h` to make the menu invisible and not include any of its images/data in the rom.

- Preresiquite fixes:
    - 
    - Improvements to fasttext (used for status repolling info display)
        - Added `dl_fasttext_begin` and `dl_fasttext_end`
        - Added `fasttext_get_str_width()`, which can be used to help center text
    - Added lerp, remap, and get_cycle functions (used for status repolling info display)
    - Fixed some segment2 externs

- Remaining issues (not urgent, can be fixed in a later PR):
    - 
    - Rumble Pak graphic translations won't work until multilang PR is merged
    - No accessory type detection (other than rumble pak) yet
    - No translations for controller select menu instructions yet
    - The controller select menu might later be changed to wait until all inputs are released before exiting after assigning all players
    - Input display on controllers doesn't include analog sticks
    - `check_repoll_gcn_origins` runs the same check for each controller that `__osPackReadEx` does. Is there a better way to not run the same check twice that doesn't make `__osPackReadEx` super messy, or is it fine as is?
    - EEPROM and SRAM need testing

### CODE REVIEW

**Comment by arthurtilly** [comment]
i'm very sus on this PR personally, wiseguy has said many times that polling controllers is a big source of lag, and this seemingly doubles the amount of polling per frame

**Review by gheskett — COMMENTED** [review]
The code itself seems fine to me, but I'm gonna refrain from approving until we have numbers that prove this won't introduce lag or anything that comes as a result of additional polling.

Additionally, I see this potentially causing issues for the GC controller branch so it also makes me feel a little hesitant to approve from that standpoint. Realistically console is going to be unsustainable with more than 2 when applied to the context of SM64 multiplayer, so maybe it'd be best to lock that behind emu for 3+? (Anybody capable of making a 4P game is going to be smart enough to figure out how to add controller support, which is another reason I hesitate to approve this if it has a non-negligible effect on runtime speed.)

**Comment by Arceveti** [comment]
> i'm very sus on this PR personally, wiseguy has said many times that polling controllers is a big source of lag, and this seemingly doubles the amount of polling per frame

Vanilla always polls all 4 ports. This PR now actually stops the polling on ports after the last used one.

**Comment by Reonu** [comment]
The most common usecase for gamecube controllers is to have it on port 2 and use only that for playing the game. So for a singleplayer hack, if there's a GC controller on port 2 and an N64 controller on port 1, there's no need to poll port 1. 

Would it be possible to account for this? Have a way to only poll port 2 in this situation?

**Comment by thecozies** [comment]
@Reonu unfortunately you cannot, it may be a limitation of libultra itself which would require a change on that end of it's even possible

**Review by Reonu — APPROVED** [review]
well given that my only concern seems to be impossible to address, I'm ok with this

File: src/game/game_init.c
Comment (thecozies) [inline src/game/game_init.c:699]: this should also have `MIN` with `MAX_ALLOWED_CONTROLLERS` right? it's a weird situation but clamping to usable controllers makes sense if somebody has controllers plugged into the first and last port for no reason. we shouldn't open up port 3 or 4 unless max controllers exceeds it

**Review by thecozies — CHANGES_REQUESTED** [review]

File: src/game/game_init.c
Comment (thecozies) [inline src/game/game_init.c:653]: another thing, what happens if a controller isn't plugged in at all? this should default to 0 to be safe so that this gets initialized properly

**Comment by FazanaJ** [comment]
arbitrary controller port reading will be possible when sourced libultra is here

**Comment by Reonu** [comment]
in any case this can be merged now and then arbitrary port reading can be implemented later imo

File: src/game/game_init.c
Comment (Arceveti) [reply src/game/game_init.c:699]: That's already covered by the loop in that function, with `cont < MAX_ALLOWED_CONTROLLERS`. Having it in `osContSetCh` with a `MIN` caused issues because it uses port numbers and `MAX_ALLOWED_CONTROLLERS ` is number of controllers. For example, having one controller plugged into port 3 or 4 when `MAX_ALLOWED_CONTROLLERS` is 2 wouldn't work.

Comment (thecozies) [reply src/game/game_init.c:699]: I don't think that's a good idea considering any port higher than 2 would cause a good degree of lag. I think it's better in that case to say that the "controller isn't connected into port 1", and then later we could consider adding a hot swap feature if the controller isnt plugged into port 1

I just think it's not necessary to allow any other ports beyond the max for super mild convenience if it has such a big drawback

Comment (gheskett) [reply src/game/game_init.c:699]: My understanding was this would only happen if you had controllers plugged into those ports, which on one hand having something with a few hundred μs extra lag is better than having something completely nonfunctional. That said, people that have controllers plugged into those ports simply out of laziness or cluelessness will experience additional lag for no good reason.

Comment (Arceveti) [reply src/game/game_init.c:699]: Being able to use any port for player 1 is vanilla behavior, I don't think we should remove that entirely. Maybe that can be a separate config define?

Comment (thecozies) [reply src/game/game_init.c:699]: any port for player 1 is vanilla? I did not realize that actually. but yeah after the first controller is connected I think that the number of supported players should dictate the maximum port. like if you are plugged into 1 and 4, just poll 1 if you only support single player

Comment (Reonu) [reply src/game/game_init.c:699]: I agree wtih cozies

Comment (Arceveti) [reply src/game/game_init.c:699]: That's what the define would do if you set it to 1, but it needs to be set to 2 for the ED64p thing.

Should the ED64p check bypass the define and ignore the N64 controller completely and only use the first GameCube controller as Player 1 instead of swapping them, if the define is set to 1?

File: src/game/game_init.c
Comment (Arceveti) [reply src/game/game_init.c:659]: It's initialized in the loop.

**Comment by Arceveti** [comment]
> arbitrary controller port reading will be possible when sourced libultra is here

Already done actually :)

File: src/game/game_init.c
Comment (thecozies) [inline src/game/game_init.c:701]: this is great, but a thought for later, i think we should have some prints under different log levels

**Comment by gheskett** [comment]
Moving to 2.2/3.0.

File: bin/segment2.c
Comment (arthurtilly) [inline bin/segment2.c:2973]: these should be .inc.c files for consistency with the repo. i can help you create a makefile rule to exclude them from the C file list if you like (this is something that should be done anyway tbh)

File: include/config/config_goddard.h
Comment (arthurtilly) [inline include/config/config_goddard.h:10]: whys this being toggled?

File: include/config/config_rom.h
Comment (arthurtilly) [inline include/config/config_rom.h:48]: why are you adding another 3 tiny random defines that literally no-one is going to know what they do. this is awful lol we're trying to avoid exactly this

File: include/ique/PR/os_cont.h
Comment (arthurtilly) [inline include/ique/PR/os_cont.h:85]: you might wanna remove these comments, also maybe wait until the math_util pr is merged but yeah just include that file

File: include/sm64.h
Comment (arthurtilly) [inline include/sm64.h:716]: why is this being set to one of the gamecube buttons? seems a bit weird if its meant to be unused by everything except the demo inputs

File: include/types.h
Comment (arthurtilly) [inline include/types.h:28]: i don't think the packed attribute is doing much here

File: include/types.h
Comment (arthurtilly) [inline include/types.h:52]: these are just formatting changes? the 0x04 stuff is not really relevant for hacking, they are just a remnant of decomp. idk why youd go out of your way to add more of them in

File: src/engine/math_util.h
Comment (arthurtilly) [inline src/engine/math_util.h:502]: a lot of these are going to conflict with the math utils PR and i am not a fan of them. is a single one of these actually being used or are you just trying to add even more macros for the sake of it

Comment (arthurtilly) [reply src/engine/math_util.h:502]: actually i didnt notice that they're just being moved from below, but regardless i think these are already gone and idk why you bothered to just move them

File: src/game/camera.c
Comment (arthurtilly) [inline src/game/camera.c:2631]: im really not a fan of this approach given how most people are used to using gPlayer1Controller. if anything gPlayer1Controller should be turned into a define and made to always point to mario's controller to make sure these code changes don't have to be made

**Review by arthurtilly — CHANGES_REQUESTED** [review]
there is a lot of weird shit in this PR that is going to need to be discussed a lot more before it can be approved. additionally this seems to have a bunch of conflicts with both my math utils PR and the existing fix rumble PR so discussion should probably wait until those are merged into this branch. i've marked the things that jump out to me in the meantime

**Comment by arthurtilly** [comment]
This pr should be repointed to 3.0 in my opinion.

File: include/config/config_goddard.h
Comment (Arceveti) [reply include/config/config_goddard.h:10]: Temporarily for testing the rumble pak graphic. I'll re-disable it once this is ready to merge.

File: src/engine/math_util.h
Comment (Arceveti) [reply src/engine/math_util.h]: This would have been done before the math util cleanup PR. These are moved back now.

File: include/types.h
Comment (Arceveti) [reply include/types.h]: Reverted

File: include/types.h
Comment (Arceveti) [reply include/types.h]: Reverted

File: include/ique/PR/os_cont.h
Comment (Arceveti) [reply include/ique/PR/os_cont.h]: I moved these macros to joybus.h instead.

**Pull request closed without merging**
